Add network stats for Windows nodes and containers

feiskyer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Add network stats for Windows nodes and containers.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #74101

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Add network stats for Windows nodes and containers
```

/sig windows
/sig node
/assign @yujuhong @PatrickLang @michmike

feiskyer commented:
/milestone v1.14

michmike commented:
@benmoss, @aserdean  can you PTAL

michmike commented on pkg/kubelet/stats/BUILD:
```diff
@@ -5,6 +5,8 @@ go_library(
     srcs = [
         "cadvisor_stats_provider.go",
         "cri_stats_provider.go",
+        "cri_stats_provider_unsupported.go",
```
what does it mean for the provider to be unsupported?

feiskyer replied:
it means some interfaces of criStatsProvider would return nil. Oh, let me rename it to others for clear.

michmike commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -158,7 +163,7 @@ func (p *criStatsProvider) ListPodStats() ([]statsapi.PodStats, error) {
 		// Fill available stats for full set of required pod stats
 		cs := p.makeContainerStats(stats, container, &rootFsInfo, fsIDtoInfo, podSandbox.GetMetadata().GetUid())
-		p.addPodNetworkStats(ps, podSandboxID, caInfos, cs)
+		p.addPodNetworkStats(ps, podSandboxID, caInfos, cs, containerNetworkStats[podSandboxID])
```
is it possible a race condition exists where a new container was created but you don't have stats for it in the dictionary?

feiskyer replied:
yep, it's possible. In that case, the network stats is actually nil, so it would just return without doing anything.

michmike commented on pkg/kubelet/winstats/network_stats.go:
```diff
@@ -0,0 +1,312 @@
+		adapters = append(adapters, adapterName)
+		newStat := n.adapterStats[adapterName]
+		newStat.Name = adapterName
+		newStat.TxPackets = newStat.TxPackets + value
```
shouldn't this be newStat.RxPackets?

feiskyer replied:
Good catch, stupid typo here.

michmike commented:
looks good overall, even though there is a lot of duplication of code that would have been better in an abstraction. i noted a small bug in there.

have you tested this and how does performance relate to the previous code?
thanks for all the work!

feiskyer commented:
Fixed the bug now. PTAL.

> have you tested this and how does performance relate to the previous code?

Yep, kubelet consumes much less CPU now (kubelet 1-4% from my cluster).

feiskyer commented:
/retest

benmoss commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -368,11 +373,21 @@ func (p *criStatsProvider) addPodNetworkStats(
+
+	// Not found from cadvisor, get from netStats.
+	if netStats != nil {
+		ps.Network = netStats
```
Maybe the log line on 395 should be updated to reflect that it is not just cadvisor stats? "Unable to find network stats for sandbox"?

feiskyer replied:
Good catch. Would update the log message.

benmoss commented on pkg/kubelet/stats/cri_stats_provider_others.go:
```diff
@@ -0,0 +1,29 @@
+// listContainerNetworkStats returns the network stats of all the running containers.
+func (p *criStatsProvider) listContainerNetworkStats() (map[string]*statsapi.NetworkStats, error) {
+	// Always return nil for unsupported platforms.
+	return nil, nil
```
This is going to yield a nil pointer exception on unsupported platforms on cri_stats_provider.go:166. Not sure if it matters, but maybe this could return an error like https://github.com/kubernetes/kubernetes/blob/48def74e641f8dff50fcf7c33c2a412593e2fede/pkg/util/mount/exec_mount_unsupported.go ?

feiskyer replied:
In such case, we still want to continue other stats logic in callers. So nil is returned here instead of errors.

feiskyer replied:
For Linx, there always have network stats, but for Windows, there doesn't before this PR.  After this, both Windows and Linux containers would have network stats.

benmoss commented on pkg/kubelet/winstats/network_stats.go:
```diff
@@ -0,0 +1,312 @@
+}
+
+func (n *networkCounter) mergePacketsReceivedPerSecondData(packetsReceivedPerSecondData map[string]uint64) []string {
+	adapters := make([]string, 0)
```
Nitpicky but I'd say either use `make([]string, 0, len(packetsReceivedPerSecondData)` or just `var adapters []string`

feiskyer replied:
Good catch. Fixed in the new commit

feiskyer commented:
/retest

michmike commented:
/approve

michmike commented:
/lgtm

feiskyer commented:
/retest

benmoss commented:
/hold 

I'm seeing a bug where the same container interface is showing up several times in metrics. It's also not reporting the name of the interface correctly, it should be `b3146b7c1f7558e9181c30ca7e155909e761726f17b581d34a5bd39b22dcf8c8_cbr0` instead of the endpoint ID.

```
{
  "node": {
    "nodeName": "34bf5e79-8af6-4a20-bb5b-d96f6d8739f0",
    "systemContainers": [
      {
        "name": "pods",
        "startTime": "2019-03-05T15:23:42Z",
        "cpu": {
          "time": "2019-03-05T15:23:42Z",
          "usageCoreNanoSeconds": 11968750000
        },
        "memory": {
          "time": "2019-03-05T15:23:42Z",
          "workingSetBytes": 86847488
        },
        "userDefinedMetrics": null
      }
    ],
    "startTime": "2019-03-04T21:20:47Z",
    "cpu": {
      "time": "2019-03-05T15:23:40Z",
      "usageCoreNanoSeconds": 159440000000
    },
    "memory": {
      "time": "2019-03-05T15:23:40Z",
      "availableBytes": 15842922496,
      "usageBytes": 3449454592,
      "workingSetBytes": 1336410112,
      "rssBytes": 0,
      "pageFaults": 0,
      "majorPageFaults": 0
    },
    "network": {
      "time": "2019-03-05T15:23:40Z",
      "name": "",
      "interfaces": [
        {
          "name": "Hyper-V Virtual Ethernet Adapter _3",
          "rxBytes": 13685,
          "rxErrors": 0,
          "txBytes": 4126,
          "txErrors": 0
        },
        {
          "name": "Hyper-V Virtual Switch Extension Adapter _2",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 0,
          "txErrors": 0
        },
        {
          "name": "Hyper-V Virtual Switch Extension Adapter",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 0,
          "txErrors": 0
        },
        {
          "name": "Teredo Tunneling Pseudo-Interface",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 0,
          "txErrors": 0
        },
        {
          "name": "6to4 Adapter",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 0,
          "txErrors": 0
        },
        {
          "name": "Hyper-V Virtual Ethernet Adapter",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 1124,
          "txErrors": 0
        },
        {
          "name": "vmxnet3 Ethernet Adapter",
          "rxBytes": 2692150,
          "rxErrors": 0,
          "txBytes": 2154344,
          "txErrors": 0
        },
        {
          "name": "Hyper-V Virtual Ethernet Adapter _2",
          "rxBytes": 2130509,
          "rxErrors": 0,
          "txBytes": 2382397,
          "txErrors": 0
        },
        {
          "name": "Microsoft IP-HTTPS Platform Interface",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 0,
          "txErrors": 0
        },
        {
          "name": "Microsoft Kernel Debug Network Adapter",
          "rxBytes": 0,
          "rxErrors": 0,
          "txBytes": 0,
          "txErrors": 0
        }
      ]
    },
    "fs": {
      "time": "2019-03-05T15:23:40Z",
      "availableBytes": 49568579584,
      "capacityBytes": 64422408192,
      "usedBytes": 14853828608
    },
    "runtime": {
      "imageFs": {
        "time": "2019-03-05T15:23:40Z",
        "availableBytes": 49568579584,
        "capacityBytes": 64422408192,
        "usedBytes": 14853828608
      }
    }
  },
  "pods": [
    {
      "podRef": {
        "name": "win-webserver-69f47cc6b6-965rt",
        "namespace": "default",
        "uid": "86a8329a-3f58-11e9-b270-005056ba5a6a"
      },
      "startTime": "2019-03-05T15:08:52Z",
      "containers": [
        {
          "name": "windowswebserver",
          "startTime": "2019-03-05T15:08:53Z",
          "cpu": {
            "time": "2019-03-05T15:23:42Z",
            "usageNanoCores": 0,
            "usageCoreNanoSeconds": 11968750000
          },
          "memory": {
            "time": "2019-03-05T15:23:42Z",
            "workingSetBytes": 86847488
          },
          "rootfs": {
            "time": "2019-03-05T15:23:42Z",
            "availableBytes": 49568579584,
            "capacityBytes": 64422408192,
            "usedBytes": 0
          },
          "logs": {
            "time": "2019-03-05T15:23:40Z",
            "availableBytes": 49568579584,
            "capacityBytes": 64422408192,
            "usedBytes": 14853828608,
            "inodesUsed": 0
          },
          "userDefinedMetrics": null
        }
      ],
      "cpu": {
        "time": "2019-03-05T15:08:53Z",
        "usageNanoCores": 0,
        "usageCoreNanoSeconds": 11968750000
      },
      "memory": {
        "time": "2019-03-05T15:23:42Z",
        "availableBytes": 0,
        "usageBytes": 0,
        "workingSetBytes": 86847488,
        "rssBytes": 0,
        "pageFaults": 0,
        "majorPageFaults": 0
      },
      "network": {
        "time": "2019-03-05T15:23:42Z",
        "name": "E4072093-AED5-4DD5-9646-0B35014E36C6",
        "rxBytes": 192356,
        "txBytes": 284827,
        "interfaces": [
          {
            "name": "E4072093-AED5-4DD5-9646-0B35014E36C6",
            "rxBytes": 192356,
            "txBytes": 284827
          },
          {
            "name": "E4072093-AED5-4DD5-9646-0B35014E36C6",
            "rxBytes": 192356,
            "txBytes": 284827
          },
          {
            "name": "E4072093-AED5-4DD5-9646-0B35014E36C6",
            "rxBytes": 192356,
            "txBytes": 284827
          },
          {
            "name": "E4072093-AED5-4DD5-9646-0B35014E36C6",
            "rxBytes": 192356,
            "txBytes": 284827
          },
          {
            "name": "E4072093-AED5-4DD5-9646-0B35014E36C6",
            "rxBytes": 192356,
            "txBytes": 284827
          },
          {
            "name": "E4072093-AED5-4DD5-9646-0B35014E36C6",
            "rxBytes": 192356,
            "txBytes": 284827
          }
        ]
      },
      "volume": [
        {
          "time": "2019-03-05T15:12:35Z",
          "availableBytes": 49573482496,
          "capacityBytes": 64422408192,
          "usedBytes": 14848925696,
          "inodesFree": 0,
          "inodes": 0,
          "inodesUsed": 0,
          "name": "default-token-kj659"
        }
      ],
      "ephemeral-storage": {
        "time": "2019-03-05T15:23:42Z",
        "availableBytes": 49568579584,
        "capacityBytes": 64422408192,
        "usedBytes": 14853828608
      }
    }
  ]
}
```

adelina-t commented:
/cc @adelina-t

PatrickLang commented:
`kubectl proxy`
`curl -s localhost:8001/api/v1/nodes/2614k8s000/proxy/stats/summary`

shows multiple entries for me too:
```
…
"network": {
     "time": "2019-03-06T01:47:25Z",
     "name": "6AFFB263-7EF5-4DE2-A24B-0C5E9156003F",
     "rxBytes": 255974,
     "txBytes": 571769,
     "interfaces": [
      {
       "name": "6AFFB263-7EF5-4DE2-A24B-0C5E9156003F",
       "rxBytes": 255974,
       "txBytes": 571769
      },
      {
       "name": "6AFFB263-7EF5-4DE2-A24B-0C5E9156003F",
       "rxBytes": 255974,
       "txBytes": 571769
      },
      {
       "name": "6AFFB263-7EF5-4DE2-A24B-0C5E9156003F",
       "rxBytes": 255974,
       "txBytes": 571769
      },
      {
       "name": "6AFFB263-7EF5-4DE2-A24B-0C5E9156003F",
       "rxBytes": 255974,
       "txBytes": 571769
      }
     ]
    },
…
```

PatrickLang commented:
@dineshgovindasamy @madhanrm - any ideas here? 

If I run a pod with 1 container- I get 4 interfaces returned from the stats API

If I run a pod with 2 containers - I get 6 interfaces returned from the stats API

feiskyer commented:
@benmoss @PatrickLang Which version of Windows are you running? I'm running with `10.0.17763.316` and each container is getting exactly one interface.

michmike commented:
@feiskyer do you have access to a host with this KB on it? https://support.microsoft.com/en-us/help/4482887

I wonder if it will make a difference. also @benmoss talked about the interface name being incorrect. do you see that issue at all?

feiskyer commented:
Fixed interface names and removed the duplicated interfaces in the new commit.

After this commit:

```json
{
  "time": "2019-03-06T06:12:34Z",
  "name": "81b01c374c9e6068fc916ce9cf07a101033f7eef7959664feb6fad7c9a2532c9_l2bridge",
  "rxBytes": 122928,
  "txBytes": 23619,
  "interfaces": [
    {
      "name": "81b01c374c9e6068fc916ce9cf07a101033f7eef7959664feb6fad7c9a2532c9_l2bridge",
      "rxBytes": 122928,
      "txBytes": 23619
    }
  ]
}
{
  "time": "2019-03-06T06:12:34Z",
  "name": "5c67dd6c016799cbba7b852e6f9be9ff5336d9be49bf71ad43393a654946ce52_l2bridge",
  "rxBytes": 219711,
  "txBytes": 27640,
  "interfaces": [
    {
      "name": "5c67dd6c016799cbba7b852e6f9be9ff5336d9be49bf71ad43393a654946ce52_l2bridge",
      "rxBytes": 219711,
      "txBytes": 27640
    }
  ]
}
{
  "time": "2019-03-06T06:12:34Z",
  "name": "42e644d1d79661aae802fd14b1b2b42a92bdd69536028b66ec401c93869722af_l2bridge",
  "rxBytes": 212104,
  "txBytes": 28040,
  "interfaces": [
    {
      "name": "42e644d1d79661aae802fd14b1b2b42a92bdd69536028b66ec401c93869722af_l2bridge",
      "rxBytes": 212104,
      "txBytes": 28040
    }
  ]
}
```

feiskyer commented:
@PatrickLang @benmoss @michmike Could you validate the new commits on your Node and check whether this is working. I wonder the duplicate interface issues are caused by different Windows versions.

michmike commented:
/retest

feiskyer commented:
/retest

benmoss commented:
/lgtm 
/hold cancel

michmike commented:
/lgtm

michmike commented:
/approve

michmike commented:
/assign @yujuhong

yujuhong commented:
/hold
Reviewing and I will some questions (will post soon). Holding this PR to avoid it from being merged accidentally.

yujuhong commented:
/test pull-kubernetes-e2e-containerd-gce

PatrickLang commented:
I'm only seeing one network now.

```
    "network": {
      "time": "2019-03-06T19:42:34Z",
      "name": "e76e4a10-eth0",
      "rxBytes": 146563,
      "txBytes": 491768,
      "interfaces": [
        {
          "name": "e76e4a10-eth0",
          "rxBytes": 146563,
          "txBytes": 491768
        }
      ]
    },
```

Looks better :)

yujuhong commented on pkg/kubelet/winstats/network_stats.go:
```diff
@@ -0,0 +1,312 @@
+}
+
+func (n *networkCounter) listInterfaceStats() []cadvisorapi.InterfaceStats {
+	stats := make([]cadvisorapi.InterfaceStats, 0)
```
nit: `s/0/len(n.adapterStats)`

yujuhong commented on pkg/kubelet/winstats/perfcounter_nodestats.go:
```diff
@@ -101,8 +101,13 @@ func (p *perfCounterNodeStatsClient) startMonitoring() error {
+
 	go wait.Forever(func() {
-		p.collectMetricsData(cpuCounter, memWorkingSetCounter, memCommittedBytesCounter)
+		p.collectMetricsData(cpuCounter, memWorkingSetCounter, memCommittedBytesCounter, networkAdapterCounter)
```
Is the `perfCounterUpdatePeriod` (1s) still reasonable after adding more metrics?

feiskyer replied:
`1s` should still work. And for network stats, there are some metrics per second. To ensure the cumulative data correctly, 1s should be used.

yujuhong reviewed (commented):
The CRI stats path is not tested by dockershim+linux AFAIK. I can't get the presubmit containerd job to run right now, so we'll need to check if post-submit CI will be affected by this change.

Other than that, this mostly looks good with questions/comments. 

The summary api test is in the node e2e suite (cannot be run by windows yet). I think we should port the test over to a regular e2e, and do some sanity check to see what metrics are present. I don't think it should block this PR, but that should be addressed afterwards.

PatrickLang commented:
/retest pull-kubernetes-e2e-containerd-gce

feiskyer commented on pkg/kubelet/stats/cri_stats_provider_windows.go:
```diff
@@ -0,0 +1,91 @@
+	for _, c := range containers {
+		container, err := hcsshim.OpenContainer(c.ID)
+		if err != nil {
+			klog.Warningf("Failed to open container %q with error '%v', continue to get stats for other containers", c.ID, err)
```
e.g. the container is listed in above list, but it's shutdown/deleted later here. Good catch, v(4) is more reasonable.

feiskyer commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -161,6 +160,12 @@ func (p *criStatsProvider) listPodStats(updateCPUNanoCoreUsage bool) ([]statsapi
 	}
 	caInfos := getCRICadvisorStats(allInfos)
 
+	// get network stats for containers.
```
sure

feiskyer replied:
sure

yujuhong commented:
> pull-kubernetes-e2e-containerd-gce

We haven't triggered this job for a while. Seems like it doesn't work anymore. We'll watch the post-submit CI instead.

feiskyer commented:
> The summary api test is in the node e2e suite (cannot be run by windows yet). I think we should port the test over to a regular e2e, and do some sanity check to see what metrics are present. I don't think it should block this PR, but that should be addressed afterwards.

@PatrickLang Could you open an issue for tracking this? We should enable node e2e tests for Windows.

feiskyer commented on pkg/kubelet/stats/helper.go:
```diff
@@ -158,6 +157,10 @@ func cadvisorInfoToNetworkStats(name string, info *cadvisorapiv2.ContainerInfo)
 		return nil
 	}
 
+	if cstat.Network == nil {
```
For Windows, there's no network network stats. This condition is used in `addPodNetworkStats()`, which gets the network stats from new param `netStats *statsapi.NetworkStats`

feiskyer committed:
Add network stats for Windows containers

feiskyer committed:
Add network stats for node interfaces

feiskyer committed:
Run hack/update-bazel.sh

PatrickLang commented:
/LGTM
@yujuhong can you approve?

yujuhong commented:
/hold cancel
/lgtm
/approve
